digdag-core/src/main/java/io/digdag/core/database/DatabaseSecretControlStore.java

```diff
@@ -61,9 +72,29 @@ public void deleteProjectSecret(int projectId, String scope, String key)
                 " where site_id = :siteId and project_id = :projectId and scope = :scope and key = :key")
         int deleteProjectSecret(@Bind("siteId") int siteId, @Bind("projectId") int projectId, @Bind("scope") String scope, @Bind("key") String key);
 
+        int upsertProjectSecret(int siteId, int projectId, String scope, String key, String engine, String value);
+    }
+
+    interface PgDao
+            extends Dao
+    {
+        @Override
         @SqlUpdate("insert into secrets" +
                 " (site_id, project_id, scope, key, engine, value, updated_at)" +
-                " values (:siteId, :projectId, :scope, :key, :engine, :value, now())")
-        int insertProjectSecret(@Bind("siteId") int siteId, @Bind("projectId") int projectId, @Bind("scope") String scope, @Bind("key") String key, @Bind("engine") String engine, @Bind("value") String value);
+                " values (:siteId, :projectId, :scope, :key, :engine, :value, now())" +
+                " on conflict (site_id, project_id, scope, key) do update set engine = :engine, value = :value, updated_at = now()")
```

Comment:
INSERT ~ ON CONFLICT is supported by PostgreSQL 9.5 or newer. I'm not sure if we support PostgreSQL 9.4 or older as Digdag-server's database...